kafka_checked/src/main.rs

```diff
@@ -0,0 +1,339 @@
+    #[serde(rename = "@val")]
+    val: i32,
+    obstime: Vec<Obstime>,
+}
+#[derive(Debug, Deserialize)]
+/// Represents <obstime>...</obstime>
+struct Obstime {
+    #[serde(rename = "@val")]
+    val: String, // avoiding parsing time at this point...
+    tbtime: Vec<Tbtime>,
+}
+#[derive(Debug, Deserialize)]
+/// Represents <tbtime>...</tbtime>
+struct Tbtime {
+    #[serde(rename = "@val")]
+    val: String, // avoiding parsing time at this point...
+    sensor: Vec<Sensor>,
+    _kvtextdata: Option<Vec<Kvtextdata>>,
+}
+#[derive(Debug, Deserialize)]
+/// Represents <sensor>...</sensor>
+struct Sensor {
+    #[serde(rename = "@val")]
+    val: Option<i32>,
+    level: Vec<Level>,
+}
+/// Represents <level>...</level>
+#[derive(Debug, Deserialize)]
+struct Level {
+    #[serde(rename = "@val")]
+    val: Option<i32>,
+    kvdata: Option<Vec<Kvdata>>,
+}
+/// Represents <kvdata>...</kvdata>
+#[derive(Debug, Deserialize)]
+struct Kvdata {
+    #[serde(rename = "@paramid")]
+    paramid: i32,
+    original: Option<f64>,
+    corrected: Option<f64>,
+    controlinfo: Option<String>,
+    useinfo: Option<String>,
+    cfailed: Option<String>,
+}
+/// Represents <kvtextdata>...</kvtextdata>
+#[derive(Debug, Deserialize)]
+struct Kvtextdata {
+    _paramid: Option<i32>,
+    _original: Option<String>,
+}
+#[derive(Debug, Deserialize)]
+struct KvalobsId {
+    station: i32,
+    paramid: i32,
+    typeid: i32,
+    sensor: Option<i32>,
+    level: Option<i32>,
+}
+
+#[derive(Debug)]
+enum Command {
+    Write {
+        kvid: KvalobsId,
+        obstime: chrono::NaiveDateTime,
+        kvdata: Kvdata,
+    },
+}
+
+#[tokio::main]
+async fn main() {
+    // parse args
+    let args: Vec<String> = env::args().collect();
+
+    assert!(args.len() >= 5, "Not enough args passed in, at least group name (for kafka), host, user, dbname needed, and optionally password (for the database)");
+
+    let mut connect_string = format!("host={} user={} dbname={}", &args[2], &args[3], &args[4]);
+    if args.len() > 5 {
+        connect_string.push_str(" password=");
+        connect_string.push_str(&args[5]);
+    }
+    println!(
+        "Connecting to database with string: {:?} \n",
+        connect_string
+    );
+
+    let group_string = format!("{}", &args[1]);
+    println!(
+        "Creating kafka consumer with group name: {:?} \n",
+        group_string
+    );
+
+    // create a channel
+    let (tx, mut rx) = mpsc::channel(10);
+
+    // start read kafka task
+    tokio::spawn(async move {
+        read_kafka(group_string, tx).await;
+    });
+
+    // Connect to the database.
+    let (client, connection) = tokio_postgres::connect(connect_string.as_str(), NoTls)
+        .await
+        .unwrap();
+    tokio::spawn(async move {
+        if let Err(e) = connection.await {
+            eprintln!("Connection error: {e}");
+        }
+    });
+
+    // Start receiving messages
+    while let Some(cmd) = rx.recv().await {
+        match cmd {
```

Comment:
Hmm, you're matching on an enum variant here, but are we expecting do send anything to the db task other than writes?

Reply:
Simplified, since likely not.